### PR TITLE
restore: support restore from s3 

### LINKED ISSUE #69: Restore from S3 compatible API?
A feature request for your roadmap:

Can it be possible to restore directly from a mydumper backup stored in S3?  In most cloud deployments this is where user backups will be stored (the S3 API is implemented by many other object stores).

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
Support restore data files from s3 storage
close #69

Related change in br: https://github.com/pingcap/br/pull/446

### What is changed and how it works?
Replace the direct file IO by  ExternalStorage interface and share the storage implements from br.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -40,3 +40,5 @@ require (
 	gopkg.in/mgo.v2 v2.0.0-20180705113604-9856a29383ce // indirect
 	modernc.org/mathutil v1.0.0
 )
+
+replace github.com/pingcap/br => github.com/glorv/br v0.0.0-20200730065835-fb34eff52143
```
Comment (kennytm) [inline go.mod:44]: 🤔 is there a corresponding BR PR?

File: lightning/config/config.go [hunk 2]
```diff
@@ -61,7 +64,10 @@ const (
 	ErrorOnDup = "error"
 )
 
-var defaultConfigPaths = []string{"tidb-lightning.toml", "conf/tidb-lightning.toml"}
+var (
+	defaultConfigPaths    = []string{"tidb-lightning.toml", "conf/tidb-lightning.toml"}
+	supportedStorageTypes = []string{"file", "local", "s3"}
```
Comment (3pointer) [inline lightning/config/config.go:69]: what's the different between file and local?

Comment (glorv) [reply lightning/config/config.go:69]: I think there is no difference, `storage.ParseBackend` just treat them the same way

File: lightning/mydump/reader_test.go [hunk 3]
```diff
@@ -104,10 +123,15 @@ func (s *testMydumpReaderSuite) TestExportStatementGBK(c *C) {
 	c.Assert(err, IsNil)
 	_, err = file.Write([]byte("');\n"))
 	c.Assert(err, IsNil)
+	stat, err := file.Stat()
+	c.Assert(err, IsNil)
 	err = file.Close()
 	c.Assert(err, IsNil)
 
-	data, err := ExportStatement(file.Name(), "auto")
+	store, err := storage.NewLocalStorage(os.TempDir())
```
Comment (kennytm) [inline lightning/mydump/reader_test.go:131]: use `c.MkDir()` for all temporary directories in tests.

File: lightning/mydump/reader.go [hunk 4]
```diff
@@ -66,21 +68,17 @@ func decodeCharacterSet(data []byte, characterSet string) ([]byte, error) {
 	return data, nil
 }
 
-func ExportStatement(sqlFile string, characterSet string) ([]byte, error) {
-	fd, err := os.Open(sqlFile)
+func ExportStatement(store storage.ExternalStorage, sqlFile FileInfo, characterSet string) ([]byte, error) {
+	fd, err := store.Open(context.Background(), sqlFile.FileMeta.Path)
```
Comment (kennytm) [inline lightning/mydump/reader.go:72]: let's make this function take a `Context`

File: lightning/mydump/loader.go [hunk 5]
```diff
@@ -259,23 +269,14 @@ func (s *mdLoaderSetup) setup(dir string) error {
...
-	err := filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
-		if err != nil {
-			return errors.Trace(err)
-		}
-
-		if f == nil || f.IsDir() {
-			return nil
-		}
-
+	err := store.WalkDir(ctx, "", 0, func(path string, size int64) error {
```
Comment (kennytm) [inline lightning/mydump/loader.go:276]: let's update BR again...

**Review by kennytm — COMMENTED** [review]
Rest LGTM

File: lightning/mydump/reader_test.go [hunk 6]
```diff
@@ -35,12 +38,18 @@ func (s *testMydumpReaderSuite) TestExportStatementNoTrailingNewLine(c *C) {
 	c.Assert(err, IsNil)
 	defer os.Remove(file.Name())
 
+	store, err := storage.NewLocalStorage(os.TempDir())
```
Comment (kennytm) [inline lightning/mydump/reader_test.go:41]: (still using `os.TempDir()`)

Comment (glorv) [reply lightning/mydump/reader_test.go]: sorry for me neglect..

**Comment by kennytm** [comment]
LGTM

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**